gateway-service-impl/src/main/java/org/hypertrace/gateway/service/explore/RequestHandler.java

```diff
@@ -95,21 +92,19 @@ QueryRequest buildQueryRequest(
     builder.setFilter(
         constructQueryServiceFilter(request, requestContext, attributeMetadataProvider));
 
+    if (requestContext.hasGroupBy() && requestContext.getOffset() > 0) {
+      // providing both group by and offset doesn't provide right results in pinot
+      // throwing unsupported operation exception for this case
+      LOG.error("Query having group by along with offset is not supported : {}", request);
+      throw new UnsupportedOperationException(
+          "Query with group by along with offset is not supported " + request);
+    }
+
     // 3. Add GroupBy
     addGroupByExpressions(builder, request);
 
-    // 4. If there's no Group By, Set Limit, Offset and Order By.
-    // Otherwise, specify a large limit and track actual limit, offset and order by expression list
-    // so we can compute
-    // these once the we get the results.
-    if (requestContext
-        .hasGroupBy()) { // Will need to do the Ordering, Limit and Offset ourselves after we get
-      // the Group By Results
-      builder.setLimit(DEFAULT_QUERY_SERVICE_GROUP_BY_LIMIT);
-      requestContext.setOrderByExpressions(getRequestOrderByExpressions(request));
-    } else { // No Group By: Use Pinot's Order By, Limit and Offset
-      addSortLimitAndOffset(request, builder);
-    }
+    // 4. Add order by along with setting limit, offset
+    addSortLimitAndOffset(request, builder);
```

Comment:
According to this comment https://github.com/hypertrace/gateway-service/blob/68ec6aabac770dcb76109d063c888cadb728df2e/gateway-service-impl/src/main/java/org/hypertrace/gateway/service/explore/RequestHandlerWithSorting.java#L9 we do it in memory since at the time Pinot did not support Group By and Order By at the same time in a query. If this has changed, there's no need to keep doing it in memory.